fe/fe-core/src/main/java/com/starrocks/alter/OnlineOptimizeJobV2.java

```diff
@@ -0,0 +1,814 @@
+    private OlapTable checkAndGetTable(Database db, long tableId) throws AlterCancelException {
+        Table table = db.getTable(tableId);
+        if (table == null) {
+            throw new AlterCancelException("table: " + tableId + " does not exist in database: " + db.getFullName());
+        }
+        Preconditions.checkState(table instanceof OlapTable);
+        return (OlapTable) table;
+    }
+
+    /**
+     * runPendingJob():
+     * 1. Create all temp partitions and wait them finished.
+     * 2. Get a new transaction id, then set job's state to WAITING_TXN
+     */
+    @Override
+    protected void runPendingJob() throws AlterCancelException {
+        Preconditions.checkState(jobState == JobState.PENDING, jobState);
+
+        LOG.info("begin to send create temp partitions. job: {}", jobId);
+        Database db = GlobalStateMgr.getCurrentState().getDb(dbId);
+        if (db == null) {
+            throw new AlterCancelException("Database " + dbId + " does not exist");
+        }
+
+        if (!checkTableStable(db)) {
+            return;
+        }
+
+        if (optimizeClause == null) {
+            throw new AlterCancelException("optimize clause is null since FE restart, job: " + jobId);
+        }
+        if (optimizeClause.isTableOptimize()) {
+            allPartitionOptimized = true;
+        }
+
+        // 1. create temp partitions
+        for (int i = 0; i < optimizeClause.getSourcePartitionIds().size(); ++i) {
+            tmpPartitionIds.add(GlobalStateMgr.getCurrentState().getNextId());
+        }
+
+        long createPartitionStartTimestamp = System.currentTimeMillis();
+        OlapTable targetTable;
+        Locker locker = new Locker();
+        locker.lockDatabase(db, LockType.READ);
+        try {
+            targetTable = checkAndGetTable(db, tableId);
+        } finally {
+            locker.unLockDatabase(db, LockType.READ);
+        }
+        try {
+            PartitionUtils.createAndAddTempPartitionsForTable(db, targetTable, postfix,
+                    optimizeClause.getSourcePartitionIds(), getTmpPartitionIds(), optimizeClause.getDistributionDesc(),
+                    warehouseId);
+            LOG.debug("create temp partitions {} success. job: {}", getTmpPartitionIds(), jobId);
+        } catch (Exception e) {
+            LOG.warn("create temp partitions failed", e);
+            throw new AlterCancelException("create temp partitions failed " + e);
+        }
+        long createPartitionElapse = System.currentTimeMillis() - createPartitionStartTimestamp;
+
+        // wait previous transactions finished
+        this.jobState = JobState.WAITING_TXN;
+        this.optimizeOperation = optimizeClause.toString();
+        span.setAttribute("createPartitionElapse", createPartitionElapse);
+        span.setAttribute("watershedTxnId", this.watershedTxnId);
+        span.addEvent("setWaitingTxn");
+
+        // write edit log
+        GlobalStateMgr.getCurrentState().getEditLog().logAlterJob(this);
```

Comment:
write `WAITING_TXN` log is not necessary because `optimizeClause` is not persisted

Reply:
persist tmp partition ids so that we can clean them when job replay